Add Espressif RMT driver

joelguittet opened the pull request:
Add initial version of the Espressif RMT driver.

A sample application is created to demonstrate the usage.

This pull-request depends on an update of hal_espressif: https://github.com/zephyrproject-rtos/hal_espressif/pull/510

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/72546

joelguittet committed:
drivers: add support for espressif rmt peripheral

The Espressif RMT driver is created from the ESP-IDF version.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

joelguittet committed:
device tree: add rmt peripheral to esp32 socs

The RMT peripheral is added to the existing relevant devices.
A DTS bindings is created to support this peripheral.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

kartben commented on include/zephyr/drivers/misc/espressif_rmt/rmt_encoder.h:
```diff
@@ -0,0 +1,182 @@
+ *      - ESP_ERR_NO_MEM: Create RMT bytes encoder failed because out of memory
+ *      - ESP_FAIL: Create RMT bytes encoder failed because of other error
+ */
+esp_err_t rmt_new_bytes_encoder(const rmt_bytes_encoder_config_t *config, rmt_encoder_handle_t *ret_encoder);
```
You're writing a zephyr driver, please return zephyr error codes :)
Also use `@retval` for documenting the discrete error values

kartben commented on dts/bindings/misc/espressif,esp32-rmt.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2025 Joel Guittet (joelguittet@gmail.com)
+# SPDX-License-Identifier: Apache-2.0
+
+description: ESP32 RMT
```
Add a title, spell out the acronym according to the guidelines, and consider adding an actual description :) (bonus point for including an example of a dts snippet as well)

kartben commented on drivers/misc/espressif_rmt/rmt_common.c:
```diff
@@ -0,0 +1,253 @@
+    int channel_id = chan->channel_id;
+    uint32_t periph_src_clk_hz = 0;
+    bool clock_selection_conflict = false;
+    // check if we need to update the group clock source, group clock source is shared by all channels
```
`/*` comments please

fabiobaltieri commented:
Sure it's fair to have it in misc if you are willing to put to work down to fix it up, problem though is that if it gets updated in idf it'll be impossible to track the changes. Anyway, your call on that.

For the rest, yeah prob fair to have it in misc to be used without a specific behavior but then I don't think we should merge a sample that makes it do a functionality that we have a subsystem for if it does not use the subsystem apis. But at that point it should be pretty straightforward to implement a led_strip driver on top of it.

joelguittet commented on drivers/misc/espressif_rmt/Kconfig:
```diff
@@ -0,0 +1,21 @@
+menuconfig ESPRESSIF_RMT
+	bool "Espressif RMT driver"
+	help
+	  Enable config options for Espressif RMT driver.
```
My comment on this proposition:
- Renaming the name of the config: ok for me.
- Why defaulting to y ? We may want to just not include it in general.
- select DMA should be added later when I will integrate DMA API for sure. However, this is only for ESP32S3 so an additional condition should be added  I guess.

joelguittet replied:
@wmrsouza For renaming the config, I cannot. The CI is checking for proper order and consistency of drivers in misc directory. I should be espressif_rmt top have vendor first, and then ESPRESSIF_RMT for the config.

wmrsouza commented:
@joelguittet I think the driver could have register_rcv_callback() api instead of read()

wmrsouza commented on dts/bindings/misc/espressif,esp32-rmt-device.yaml:
```diff
@@ -0,0 +1,28 @@
+  with-dma:
+    type: boolean
+    description: |
+      If set, the driver will allocate an RMT channel with DMA capability.
```
No need for this property. You can check if dmas is present in device tree
By the way only channel 3 on esp32s3 is DMA capable

wmrsouza replied:
@joelguittet I think you are mixing up the information regarding rmt device driver with that regarding to led_strip driver for ws2812 using rmt back end, but lets see (may be I missed something)

wmrsouza commented on dts/bindings/misc/espressif,esp32-rmt-device.yaml:
```diff
@@ -0,0 +1,28 @@
+  io-loop-back:
+    type: boolean
+    description: |
+      The signal output from the GPIO will be fed to the input path as well.
```
In my humble opinion, you should separate rmt driver PR from led_strip rmt backend PR, provide a test (with loopback for the RMT driver; with and without DMA) andd after that follow with the led_strip rmt backend PR

joelguittet committed:
boards: espressif: indicate rmt support for all relevant boards

Indicate rmt support for all relevant Espressif boards.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

joelguittet commented:
@wmrsouza 'rmt' added to all relevant espressif boards.
For the default pinctrl I don't know. How many channels ? Which GPIO to select ? There is no preferences to provide I guess. I see also the default pinctrl is for uart, spi, i2s. The other peripherals have no default pinctrl or it's somewhere else ?

wmrsouza commented:
> @wmrsouza 'rmt' added to all relevant espressif boards. For the default pinctrl I don't know. How many channels ? Which GPIO to select ? There is no preferences to provide I guess. I see also the default pinctrl is for uart, spi, i2s. The other peripherals have no default pinctrl or it's somewhere else ?

- at leat one channel 
- preferencialy GPIO near each other, and not used by other peripherals or overlaping with GPIO used by peripheral with `status = disabled;`

joelguittet commented:
> at leat one channel
preferencialy GPIO near each other, and not used by other peripherals or overlaping with GPIO used by peripheral with status = disabled;

So this is just dead code at the end? @wmrsouza

wmrsouza commented:
> > at leat one channel
> > preferencialy GPIO near each other, and not used by other peripherals or overlaping with GPIO used by peripheral with status = disabled;
> 
> So this is just dead code at the end? @wmrsouza

no, I'm NOT saying to make rmt `status="disabled";`. What I'm saying is if the board doesn't have "free" pins then ovelap those from peripherals having `status="disabled";`

joelguittet committed:
samples: drivers: add espressif rmt bytes encoder application

Adding this application to test espressif_rmt driver using bytes
encoder.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

joelguittet committed:
boards: espressif: add rmt default pinctrl for all relevant boards

Add definition for default rmt pinctrl on all relevant boards.
GPIO12 and GPIO13 are used by default.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

wmrsouza reviewed (commented):
@joelguittet Suggestion for commit title and message body
- title:
manifest: esp32: update hal_espressif
- message body
Update hal to cover latest features and fixes.

wmrsouza commented on samples/drivers/misc/espressif_rmt/src/main.c:
```diff
@@ -0,0 +1,329 @@
+	.signal_range_max_ns = 2500000,
+};
+
+static size_t example_encoder_encode(rmt_encoder_t *encoder,
```
Is the example prefix necessary? (nitpick)
The same question applies to all other occurrences.

wmrsouza commented:
@joelguittet Overral is not bad. In fact is very good code but would not be better treat rmt channels as nodes with addresses (channel index under rmt peripheral) under device tree and a setup function to configure channels (configurations that chan change and are not informet in device threem, rx callback function receive a device struct as a parameter, ...) or some more Zephyr's similar kind of api function (I know rmt is Espressif only peripheral) 

@sylvioalves @marekmatej @raffarost @LucasTambor let me know what you think about this. Might be I'm being too strict again?

joelguittet committed:
samples: drivers: add espressif rmt simple encoder application

Adding this application to test espressif_rmt driver using simple
encoder.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

joelguittet committed:
dt-bindings: pinctrl: add esp32 rmt pins definition

Adding RMT pinctrl for all relevant ESP32 devices.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

sylvioalves commented:
1. Would you also add a test case? `tests/drivers/misc/espressif_rmt/ with loopback TX→RX`

sylvioalves commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1321 @@
+#include <zephyr/drivers/dma/dma_esp32.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/pinctrl/pinctrl_esp32_common.h>
+#include <zephyr/logging/log.h>
```
would you move the LOG module not to be between #includes? same on `rmt_tx.c` file.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1321 @@
+			return;
+		}
+		/* delay a while, wait for DMA data going to RMT memory block */
+		esp_rom_delay_us(1);
```
Use Zephyr API instead: `k_busy_wait(1);`.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_rx.c:
```diff
@@ -0,0 +1,923 @@
+		 * data and reset the pointer
+		 */
+		rmt_ll_clear_interrupt_status(hal->regs, RMT_LL_EVENT_RX_ERROR(channel_id));
+		ESP_DRAM_LOGE("rmt", "hw buffer too small, received symbols truncated");
```
`LOG_ERR`. Check all files please.

joelguittet commented on drivers/misc/espressif_rmt/rmt.c:
```diff
@@ -0,0 +1,77 @@
+		(ESPRESSIF_RMT_DMA_CHANNEL_UNDEFINED))
+
+#define ESPRESSIF_RMT_INIT(idx)                                                     \
+	PINCTRL_DT_DEFINE(DT_NODELABEL(rmt));                                           \
```
Done

joelguittet commented on drivers/misc/espressif_rmt/Kconfig:
```diff
@@ -0,0 +1,37 @@
+
+config ESPRESSIF_RMT_INIT_PRIORITY
+	int "ESP32 RMT driver init priority"
+	default 90
```
Probably better with `config KERNEL_INIT_PRIORITY_DEVICE` (which default to 50).

joelguittet commented on dts/bindings/misc/espressif,esp32-rmt.yaml:
```diff
@@ -0,0 +1,76 @@
+    pinctrl is used to configure RMT channels and GPIO to be used.
+    When using ESP32 or ESP32S2 series, each RMT channel should be used once (eg. RMT_OUT0_GPIOxx
+    and RMT_IN0_GPIOxx can't be used at the same time, etc.). An error is displayed when trying to
+    create channels whith such configuration.
```
Good catch! Fixed.

joelguittet commented:
> 1. Would you also add a test case? `tests/drivers/misc/espressif_rmt/ with loopback TX→RX`

Examples in `samples/drivers/misc/espressif_rmt` directory implement a loopback. This will mean duplicate a sample to go in the tests ?

kartben commented on dts/bindings/misc/espressif,esp32-rmt.yaml:
```diff
@@ -0,0 +1,76 @@
+    in the RMT peripheral and configuring pinctrl to use DMA capable RMT channels
+    (eg. RMT_OUT3_GPIOxx and RMT_OUT3_GPIOxx on ESP32S3).
+
+    Here is an example ESP32 configuration with a RMT device:
+
+    &pinctrl {
+      rmt_default: rmt_default {
+        group1 {
+          pinmux = <RMT_OUT3_GPIO13>;
+        };
+
+        group2 {
+          pinmux = <RMT_IN3_GPIO12>;
+        };
+      };
+    };
+
+    rmt: rmt@60016000 {
+        compatible = "espressif,esp32-rmt";
+        #address-cells = <1>;
+        #size-cells = <0>;
+        reg = <0x60016000 0x1000>;
+        pinctrl-0 = <&rmt_default>;
+        pinctrl-names = "default";
+        dmas = <&dma 8>, <&dma 9>;
+        dma-names = "rx", "tx";
+        interrupts = <RMT_INTR_SOURCE IRQ_DEFAULT_PRIORITY 0>;
+        interrupt-parent = <&intc>;
+        clocks = <&clock ESP32_RMT_MODULE>;
+        status = "okay";
+    };
```
use `examples:` for this

kartben commented on dts/bindings/misc/espressif,esp32-rmt.yaml:
```diff
@@ -0,0 +1,76 @@
+# SPDX-FileCopyrightText: Copyright The Zephyr Project Contributors
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Espressif Remote Control Transceiver (RMT)
+
```
```suggestion
title: Espressif RMT (Remote Control Transceiver)

description: |
```

kartben commented on include/zephyr/drivers/misc/espressif_rmt/rmt.h:
```diff
@@ -0,0 +1,77 @@
+#define ZEPHYR_DRIVERS_MISC_ESPRESSIF_RMT_RMT_H_
+
+/**
+ * @brief Espressif RMT.
```
```suggestion
 * @brief Interfaces for Espressif RMT (Remote Control Transceiver).
```

raffarost commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1323 @@
+	return 0;
+}
+
+static bool IRAM_ATTR rmt_isr_handle_tx_threshold(rmt_tx_channel_t *tx_chan)
```
There are more functions in IRAM, so the total amount of memory might be significant indeed.

joelguittet commented:
Hello guys! I know there is 4.4.0 soon and scope is frozen, just a ping to remind you this stuff :-)
@raffarost any concern about the answers i provided to your above comments?
@sylvioalves how is it on your side? Additional review to do or is it almost done?

raffarost commented on drivers/misc/espressif_rmt/rmt_rx.c:
```diff
@@ -0,0 +1,1015 @@
+
+#if SOC_RMT_SUPPORT_SLEEP_RETENTION
+	if (config->flags.allow_pd != 0) {
+		rmt_create_retention_module(group);
```
I'm working on sleep retention support for a number of drivers.
I'll soon post a PR and will reference it here, for you to take a look.

raffarost replied:
Take a look at #107344 for reference about sleep retention (would be nice to follow the same style).
This PR will enable testing the feature on your driver, if you're using a C6.

raffarost commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1323 @@
+	return need_eof_marker ? 1 : 0;
+}
+
+static size_t IRAM_ATTR rmt_encode_check_result(rmt_tx_channel_t *tx_chan, rmt_tx_trans_desc_t *t)
```
I think at least in latest HAL/IDF, the function does not define IRAM_ATTR directly, as it has a linking script to do the job of reading the optimization functions in Kconfig (such as `RMT_TX_ISR_HANDLER_IN_IRAM`) and allocate accordingly. Take a look at `components/esp_driver_rmt/linker.lf`. In Zephyr we'll need the Kconfigs and the conditional statements are coded in `default.ld` liker file.

joelguittet replied:
I see this in the esp-idf driver. But for zephyr what should we do now?
I mean, other esp32 drivers already in Zephyr also use IRAM_ATTR (uart, i2c, etc)

joelguittet commented:
@sylvioalves you told me previously about tests. I looked a bit at existing tests generally. Seems API testing is a minimum to do. I should probably add this, correct? Do you have any recommendations or expectations on this side at Espressif?

sylvioalves commented on drivers/misc/espressif_rmt/rmt.c:
```diff
@@ -0,0 +1,93 @@
+#if SOC_RMT_SUPPORT_DMA
+	/* Check DMA device is ready (if used) */
+	if (config->dma_dev) {
+		if (!device_is_ready(config->dma_dev)) {
```
You could make this 1 single line: `if (config->dma_dev && !device_is_ready(config->dma_dev)) {`

joelguittet replied:
Sure, done

sylvioalves commented on drivers/misc/espressif_rmt/rmt.c:
```diff
@@ -0,0 +1,93 @@
+	/* Check DMA device is ready (if used) */
+	if (config->dma_dev) {
+		if (!device_is_ready(config->dma_dev)) {
+			LOG_ERR("Clock device not ready");
```
`DMA not ready`

joelguittet replied:
Good catch, and the above for clock is with a DMA in the log. Fixing it too.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_rx.c:
```diff
@@ -0,0 +1,1023 @@
+
+/* RMT RX Channel ID from pinmux configuration */
+#define RMT_RX_CHANNEL_ID(pinmux) \
+	ESP32_PIN_SIGI(pinmux) - ESP_RMT_SIG_IN0
```
Missing parentheses: 
`(ESP32_PIN_SIGI(pinmux) - ESP_RMT_SIG_IN0)`

joelguittet replied:
Done. And also for rmt_tx.c which has a similar definition

sylvioalves commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1328 @@
+	/* acquire group handle, increase reference count */
+	rmt_group_t *group = rmt_acquire_group_handle(config->tx_channel_array[0]->group->group_id);
+	/* sanity check */
+	assert(group);
```
PLease convert all `assert` to `__ASSERT_NO_MSG` instead.

joelguittet replied:
Done for all assert.

sylvioalves commented on drivers/misc/espressif_rmt/rmt.c:
```diff
@@ -0,0 +1,93 @@
+
+	/* Check clock device is ready */
+	if (!device_is_ready(config->clock_dev)) {
+		LOG_ERR("DMA device not ready");
```
RMT clock not ready

joelguittet replied:
Catched together with the previous log comment.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_rx.c:
```diff
@@ -0,0 +1,1023 @@
+	 * APB clock no matter what the clock source is used by the RMT channel as the "core"
+	 * clock
+	 */
+#if defined(CONFIG_IDF_TARGET_ESP32) || defined(CONFIG_IDF_TARGET_ESP32S2)
```
Please replace all CONFIG_IDF_xxx to proper Zephyr-like `CONFIG_SOC_ESPxxxx`.

joelguittet replied:
Done

sylvioalves commented on drivers/misc/espressif_rmt/rmt_common.c:
```diff
@@ -0,0 +1,452 @@
+
+typedef struct rmt_platform_t {
+	/**< Platform level mutex lock */
+	unsigned int key;
```
There is no need to global locking, this can go.

joelguittet replied:
True. Initially copied from ESP-IDF logic.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_common.c:
```diff
@@ -0,0 +1,452 @@
+	rmt_group_t *group = NULL;
+
+	/* prevent install rmt group concurrently */
+	s_platform.key = irq_lock();
```
Use local locking/unlocking instead.

joelguittet replied:
Done

sylvioalves commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1328 @@
+{
+	rmt_channel_t *channel = &tx_chan->base;
+	rmt_tx_trans_desc_t *trans_desc = NULL;
+	bool need_yield = false;
```
Zephyr does not need to yield. 
1. Remove all `need_yield` from the code
2. Replace `bool` function signatures to `void` that only returns/uses the `need_yield` flag.

joelguittet replied:
I know that, I was not sure what to do with it. So removing this need_yield.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1328 @@
+	if (t->loop_count != 0) {
+		if (unlikely(encoded_symbols > tx_chan->base.mem_block_num
+			* SOC_RMT_MEM_WORDS_PER_CHANNEL)) {
+			LOG_ERR("encoding artifacts can't exceed hw memory block for loop "
```
I think we should avoid/remove `LOG_ERR` from ISR path.

joelguittet replied:
Seems that many things are printed on ISR, that would mean much less information in case of error. Is there an alternative?

joelguittet replied:
@sylvioalves note the doc here https://docs.zephyrproject.org/latest/services/logging/index.html#default-frontend indicates "the logging API can be called in an interrupt".

sylvioalves commented on drivers/misc/espressif_rmt/rmt_encoder_simple.c:
```diff
@@ -0,0 +1,253 @@
+						 * Report this. EARLY_LOGE as we're running from an
+						 * ISR.
+						 */
+						LOG_ERR("rmt_encoder_simple: encoder callback "
```
Same here, can you remove the `LOG_ERR` from ISR path?

sylvioalves commented on drivers/misc/espressif_rmt/rmt_encoder_simple.c:
```diff
@@ -0,0 +1,253 @@
+	return 0;
+}
+
+static size_t rmt_encode_simple(rmt_encoder_t *encoder, rmt_channel_handle_t channel,
```
Add `IRAM_ATTR`.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_encoder_simple.c:
```diff
@@ -0,0 +1,253 @@
+}
+
+static size_t rmt_encode_simple(rmt_encoder_t *encoder, rmt_channel_handle_t channel,
+	const void *data, size_t data_size, rmt_encode_state_t *ret_state)
```
Would you fix those sonarqube issues?

joelguittet replied:
Do you mean the cognitive complexity? Seems acceptable according to me, it's not so much. 25 is the default from SonarQube that I generally consider to low in C. Good comments being the key to reduce the complexity, which is not considered in the calculation. Additionally there are about 1200 cognitive complexity issue in Sonar report actually, and some are much higher that this one...
Tell me if this is strictly required here. If so, I will do it.

joelguittet replied:
Note: seems it has gone on the latest analysis... (without changing anything)

raffarost commented on drivers/misc/espressif_rmt/rmt_common.c:
```diff
@@ -0,0 +1,452 @@
+				rmt_ll_enable_bus_clock(group_id, true);
+				rmt_ll_reset_register(group_id);
+			}
+#if CONFIG_PM && SOC_RMT_SUPPORT_SLEEP_RETENTION
```
We should enable sleep retention only when `CONFIG_PM_POWER_DOWN_PERIPHERAL_IN_LIGHT_SLEEP=y`. You could add to `rmt_private.h` something like:

```c
#if CONFIG_PM_POWER_DOWN_PERIPHERAL_IN_LIGHT_SLEEP && SOC_RMT_SUPPORT_SLEEP_RETENTION
#define RMT_SLEEP_RETENTION_ENABLED 1
#else
#define RMT_SLEEP_RETENTION_ENABLED 0
#endif
```

And use it elsewhere in the code for a clean read.

joelguittet replied:
@raffarost it seems `CONFIG_PM_POWER_DOWN_PERIPHERAL_IN_LIGHT_SLEEP` doesn't exist in Zephyr, but I see it several times in hal_espressif. The CI also complains here about this missing option. What is wrong there? Maybe it has been removed recently?

raffarost replied:
You're correct, on `main` this option is still not included. The PR #107344 is the one enabling its use since other changes are also needed for this hardware feature to work correctly. I'd suggest you remove anything related to PM for now, and we can add that in a follow up PR.

joelguittet replied:
@raffarost I moved the sleep retention stuff to a draft PR waiting yours to be merged so we are not blocked for the current PR: https://github.com/zephyrproject-rtos/zephyr/pull/108146

raffarost commented on drivers/misc/espressif_rmt/rmt_common.c:
```diff
@@ -0,0 +1,473 @@
+#if SOC_RMT_SUPPORT_SLEEP_RETENTION
+		sleep_retention_module_t module = rmt_reg_retention_info[group->group_id].module;
+
+		if (sleep_retention_is_module_created(module)) {
```
I just realized, maybe you removed the PM code from the driver.
If that's the case, all things related to sleep retention could be sacked as well, as it wouldn't make sense to only add those.

joelguittet commented:
@sylvioalves @raffarost several of your comments above fixed, please advise. Thanks!!

sylvioalves commented:
> @sylvioalves @raffarost several of your comments above fixed, please advise. Thanks!!

Checking again, thanks for updating.

sylvioalves commented:
> @sylvioalves you told me previously about tests. I looked a bit at existing tests generally. Seems API testing is a minimum to do. I should probably add this, correct? Do you have any recommendations or expectations on this side at Espressif?

Would be good if some `tests/drivers/misc/espressif_rmt/` could be added. Use gpio loopback to test the rmt driver and provide pass/fail test cases. That will be pulled into internal CI to verify in real board. We can work this out in another PR as well.

joelguittet commented:
> > @sylvioalves you told me previously about tests. I looked a bit at existing tests generally. Seems API testing is a minimum to do. I should probably add this, correct? Do you have any recommendations or expectations on this side at Espressif?
> 
> Would be good if some `tests/drivers/misc/espressif_rmt/` could be added. Use gpio loopback to test the rmt driver and provide pass/fail test cases. That will be pulled into internal CI to verify in real board. We can work this out in another PR as well.

Yes i would definitively have such test but maybe it's better on a separated PR. But I will try to work on it in parallel of current reviews here.

joelguittet committed:
tests: drivers: espressif_rmt: add loopback

Add loopback test application for espressif_rmt driver.

Signed-off-by: Joel Guittet <joelguittet@gmail.com>

joelguittet commented:
@sylvioalves I have added here a "loopback" test application, and executed it on my esp32s3-devkitc is fine. It's checking expected lines in the console. It's based on the bytes encoder sample. Tell me if this is suitable, this is the very first time i do that in Zephyr repo :-)

sylvioalves commented on drivers/misc/espressif_rmt/rmt_rx.c:
```diff
@@ -0,0 +1,998 @@
+	esp_err_t ret;
+	int rc;
+
+	/* Check if priority is valid */
```
I think you can check for NULL pointer before:
```c
if (!(config && ret_chan && config->resolution_hz && config->trans_queue_depth)) {
 		LOG_ERR("Invalid argument");
 		return -EINVAL;
 	}
```

sylvioalves replied:
Same thing on rmt_rx.c file.

sylvioalves commented on drivers/misc/espressif_rmt/rmt_tx.c:
```diff
@@ -0,0 +1,1300 @@
+	/* acquire group handle, increase reference count */
+	rmt_group_t *group = rmt_acquire_group_handle(config->tx_channel_array[0]->group->group_id);
+	/* sanity check */
+	__ASSERT_NO_MSG(group);
```
Perhaps in this case it should return -ENOMEM instead of asserting? What you think?
```c
	if (!group) {
		LOG_ERR("Unable to acquire RMT group");
		k_free(synchro);
		return -ENOMEM;
	}
```

sylvioalves commented on drivers/misc/espressif_rmt/rmt_common.c:
```diff
@@ -0,0 +1,380 @@
+	unsigned int key;
+	bool new_group = false;
+	rmt_group_t *group = NULL;
+
```
Please add null check as well:
```c
	if (group_id < 0 || group_id >= RMT_LL_GET(INST_NUM)) {
		return NULL;
	}
```

sylvioalves commented on drivers/misc/espressif_rmt/rmt_encoder_copy.c:
```diff
@@ -0,0 +1,145 @@
+	rmt_encode_state_t state = RMT_ENCODING_RESET;
+	size_t symbol_index = copy_encoder->last_symbol_index;
+	/* how many symbols will be copied by the encoder */
+	size_t mem_want = (data_size / 4 - symbol_index);
```
would you replace the magic 4 to ` sizeof(rmt_symbol_word_t)` instead?

sylvioalves commented on drivers/misc/espressif_rmt/rmt_encoder_simple.c:
```diff
@@ -0,0 +1,253 @@
+	encoder->ovf_buf = rmt_alloc_encoder_mem(min_chunk_size * sizeof(rmt_symbol_word_t));
+	if (!encoder->ovf_buf) {
+		LOG_ERR("Unable to allocate memory for overflow buffer");
+		if (encoder) {
```
I think this check can go. If it got here, it is not NULL and `k_free` is ok.

kartben commented on samples/drivers/misc/espressif_rmt/simple_encoder/prj.conf:
```diff
@@ -0,0 +1 @@
+# nothing here
```
this seems off, it should explicitly set `CONFIG_ESPRESSIF_RMT=y` as the subsystem shoudn't be `default y`

joelguittet commented:
@sylvioalves @kartben I have seen your comments, globally agree with you, just expect a few days delay on my side to fix them. But be sure, i'm still on it :-)

carlescufi commented:
Architecture WG meeting:

- RMT is a peripheral to generate infrared signals for remote control
- @joelguittet explains how the signal is generated: it's a multi-purpose signal generator without any byte code interpretation
- @bjarki-andreasen says there is a similar peripheral in Nordic (PWM) that can also do arbitrary binary pulse encoding
-

sylvioalves commented:
> Architecture WG meeting:
> 
> * RMT is a peripheral to generate infrared signals for remote control
> * @joelguittet explains how the signal is generated: it's a multi-purpose signal generator without any byte code interpretation
> * @bjarki-andreasen says there is a similar peripheral in Nordic (PWM) that can also do arbitrary binary pulse encoding

A little bit more info about RMT capabilities:
1. IR remote control — transmit/receive NEC, RC5, Sony SIRC, and other IR protocols
2. WS2812 / NeoPixel LEDs — generate the precise timing needed for addressable RGB LED strips (most popular use)
3. 1-Wire protocols — bit-banging DS18B20 temperature sensors, etc.
4. Stepper motor control — generating step pulses with precise timing
5. Servo PWM — though LEDC/MCPWM are more typical (already existing Espressif driver's in Zephyr)
6. Custom digital protocols — any application that needs accurate, hardware-timed pulse sequences without CPU intervention
7. Signal generation/capture — arbitrary waveforms, pulse measurement, frequency counting

bjarki-andreasen commented:
Architecture meeting notes to add to https://github.com/zephyrproject-rtos/zephyr/pull/101448#issuecomment-4489344390 (from memory right after meeting ended):
* @nashif Mentions public headers in misc contain HAL includes which is not allowed
* @sylvioalves Mentions the RMT has waveform capture capabilties. (note Nordic hardware does not have this capability)
* @bjarki-andreasen Proposes creating a driver class that groups similar hardware, though with vendor specific APIs, like we do for drivers/firmware and drivers/interrupt controllers to some extent.
* No hard objections to the above

sylvioalves commented:
> Architecture meeting notes to add to [#101448 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/101448#issuecomment-4489344390) (from memory right after meeting ended):
> 
> * @nashif Mentions public headers in misc contain HAL includes which is not allowed
> * @sylvioalves Mentions the RMT has waveform capture capabilties. (note Nordic hardware does not have this capability)
> * @bjarki-andreasen Proposes creating a driver class that groups similar hardware, though with vendor specific APIs, like we do for drivers/firmware and drivers/interrupt controllers to some extent.
> * No hard objections to the above

@carlescufi @bjarki-andreasen https://github.com/zephyrproject-rtos/zephyr/issues/109586

marekmatej commented:
I think the `lpcore` target tests can be filtered out.
